src/pool-bin/lens/BinQuoter.sol

```diff
@@ -9,222 +9,160 @@ import {BalanceDelta} from "pancake-v4-core/src/types/BalanceDelta.sol";
-        assembly ("memory-safe") {
-            revert(add(0x20, r), mload(r))
+            pathKey = params.path[i - 1];
+            (PoolKey memory poolKey, bool oneForZero) = pathKey.getPoolAndSwapDirection(outputCurrency);
+
+            swapDelta = _swap(poolKey, !oneForZero, amountOut.safeInt128(), pathKey.hookData);
+
+            amountOut = oneForZero ? uint128(-swapDelta.amount1()) : uint128(-swapDelta.amount0());
+            outputCurrency = pathKey.intermediateCurrency;
         }
+        // amountOut after the loop exits actually holds the amountIn of the trade
+        amountOut.revertQuote();
     }
 
     /// @dev quote an ExactOutput swap on a pool, then revert with the result
-    function _quoteExactOutputSingle(QuoteExactSingleParams memory params)
-        public
-        override
-        selfOnly
-        returns (bytes memory)
-    {
-        amountOutCached = params.exactAmount;
-
-        (BalanceDelta deltas, uint24 activeIdAfter) =
+    function _quoteExactOutputSingle(QuoteExactSingleParams calldata params) external selfOnly returns (bytes memory) {
+        BalanceDelta swapDelta =
             _swap(params.poolKey, params.zeroForOne, params.exactAmount.safeInt128(), params.hookData);
 
-        if (amountOutCached != 0) delete amountOutCached;
-        int128[] memory deltaAmounts = new int128[](2);
-
-        deltaAmounts[0] = deltas.amount0();
-        deltaAmounts[1] = deltas.amount1();
-
-        bytes memory result = abi.encode(deltaAmounts, activeIdAfter);
-        assembly ("memory-safe") {
-            revert(add(0x20, result), mload(result))
-        }
+        // the input delta of a swap is negative so we must flip it
+        uint256 amountIn = params.zeroForOne ? uint128(-swapDelta.amount0()) : uint128(-swapDelta.amount1());
+        amountIn.revertQuote();
     }
 
-    /// @dev Execute a swap and return the amounts delta, as well as relevant pool state
+    /// @dev Execute a swap and return the balance delta
     /// @notice if amountSpecified < 0, the swap is exactInput, otherwise exactOutput
     function _swap(PoolKey memory poolKey, bool zeroForOne, int128 amountSpecified, bytes memory hookData)
         private
-        returns (BalanceDelta deltas, uint24 activeIdAfter)
+        returns (BalanceDelta deltas)
     {
         deltas = poolManager.swap(poolKey, zeroForOne, amountSpecified, hookData);
 
-        // only exactOut case
-        if (amountOutCached != 0 && amountOutCached != uint128(zeroForOne ? deltas.amount1() : deltas.amount0())) {
-            revert InsufficientAmountOut();
+        /// @dev Check that the pool was not illiquid
+        /// even BinPool will emit BinPool__OutOfLiquidity when the pool is illiquid
+        /// We still need to apply the check in case hook contract manipulates the delta
+        int128 amountSpecifiedActual = (zeroForOne == (amountSpecified < 0)) ? deltas.amount0() : deltas.amount1();
```

Comment:
im curious, would this means basically if the pool has hook which charge a fee or take tokenIn for swap, quoter will fail?

Reply:
Not exactly, this will fail only when the hook does not respect the amountSpecified. For example, for exactInput the pool with hook require more/less input token or for exactOutput the pool with hook return more/less token.

Reply:
Revisited this: it seems amountSpecified can never be manipulated through hook i.e. delta.tokenOnSpecifiedSide() will always equal to amountSpecified for binPool. I start to thinking the check here is unnecessary again 😂

![CleanShot 2024-10-15 at 11 21 54@2x](https://github.com/user-attachments/assets/6fcb3a7a-6e65-4b67-98bd-1eeacc6a3840)




For CLPool, the reason why it's needed is because CLPool#swap (step2 in the screenshot) support partial fulfilled which could be BinPool__OutOfLiquidity for binPool.

Reply:
> Not exactly, this will fail only when the hook does not respect the amountSpecified. For example, for exactInput the pool with hook require more/less input token or for exactOutput the pool with hook return more/less token

`if (amountSpecifiedActual != amountSpecified)`

this case will be hit so long a hook take a fee or give token in `hook.afterSwap returnDelta` ?

Reply:
got it. yep feel free to remove that or keep that (no harm i gues) since the previous implementation included that check